### DISCUSSION TITLE
[Bug]: Basic questions regarding serving sites with Herd

### DISCUSSION THREAD

**Discussion by mceh** [beyondcode/herd-community#692] [Q&A]
### Platform

macOS

### Operating system version

Macos Sonoma 14.5

### System architecture

ARM64 (M1, M2, etc)

### Herd Version

1.6.0

### PHP Version

PHP 8.3

### Bug description

Hi,
I didn’t find a proper way to contact any support other than this GitHub discussion. Please point me to the right source if this is wrong here. I read all of the (very sparse) documentation about Herd.

I’m currently exploring ways to get away from MAMP Pro which is simply not maintained for a long time and seems not to be interested in adding modern versions of software MySQL etc.

Laravel Herd is recommended as a much faster local development package than MAMP Pro on the web, what is quite tempting. It seems easy enough to develop Wordpress-sites in a laravel environment. I made a quick test and it worked.

But I work with different CMS: e.g. Contao and Typo3. Both need a web-root within the installation directory. Contao asks specifically for /web as a publicly accessible folder (with the system-files installed above the /web directory). To install Contao you have to open the file /web/contao-manager.phar.php in the browser. Opening a php-file directly in a Herd-served directory just doesn't work.

After hours of trying I just don’t get this to work. I experimented with different Herd paths, subfolders, link and park-commands on the command line on different folders, subfolders and even php-files.

It must be possible to work on more complex use cases than just serving an index.php within a simple folder? Or am I wrong? I think if it’s feasible to develop Wordpress projects in Herd, why not other CMS…

What am I doing wrong? Or is Herd maybe the completely wrong tool for me?

Regards
Markus

**Comment by mpociot** [comment]
Herd has a concept of "drivers", that allows you to define (either globally or per-project) how Herd should load this project, when looking for the "root" of it.

Check out the documentation about custom drivers:
https://herd.laravel.com/docs/windows/1/extending-herd/custom-drivers

As well as this GitHub discussion: https://github.com/beyondcode/herd-community/discussions/186

**Reply by mceh** [reply]
Thanks, but I don't get it to work.
I tried adding the LocalValetDriver.php to –/Library/Application Support/Herd/config/valet/Drivers
with the contents of the documentation. I tried the version from the GitHub discussion. Also with the modification of the namespace as suggested in several comments. I restarted the services of Herd after every change, of course.

Disclaimer. I’m not really good at PHP. I just want to install/use my CMSs locally to develop websites.

My directory-structure for this test currently:
~/Herd/contaotest/web/contao-manager.phar.php

The URL opened by Herd is: http://contaotest.test
I tried
http://contaotest.test/web/
http://contaotest.test/web/contao-manager.phar.php

I get the error:
Fatal error: Uncaught Error: Class "Valet\Drivers\Custom\LocalValetDriver" not found in /Applications/Herd.app/Contents/Resources/valet/cli/Valet/Drivers/ValetDriver.php:58 Stack trace: #0 /Applications/Herd.app/Contents/Resources/valet/server.php(54): Valet\Drivers\ValetDriver::assign('/Users/markus/H...', 'contaotest', '/web/contao-man...') #1 {main} thrown in /Applications/Herd.app/Contents/Resources/valet/cli/Valet/Drivers/ValetDriver.php on line 58
